File: backend/src/inkflow/domain/models/model_readiness.py
"""ModelReadiness 领域模型 — 首启模型就绪判据的只读视图（#934 §3.1）。

领域层保持纯净：仅依赖 Pydantic v2，不感知 ORM / 框架 / 网络。

依据: specs/f60-first-run-guide/spec.md §2.1 / §3.1。
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel

ReadinessReason = Literal["ready", "no_provider", "no_chat_model", "no_key"]
"""未就绪原因（互斥，按优先级降级：no_provider > no_chat_model > no_key）。

- ``ready``: 已具备可解析 chat 模型 + 有效 key（``ready=True``）
- ``no_provider``: 注册表无任何 provider（真·全新安装）
- ``no_chat_model``: 有 provider 但无一含 chat 模型（#929 精确缺陷形态）
- ``no_key``: 有 chat 模型但对应 provider 无 key
"""


class ModelReadiness(BaseModel):
    """首启模型就绪判据（GET /api/v1/settings/model-readiness 响应体）。

    Attributes:
        ready: 是否可进入写作主流程（唯一门控判据，spec §2.1）。
        has_chat_model: 是否存在「有 key 且有 chat 模型」的 provider。
        has_embedding_model: 是否存在「有 key 且有 embedding 模型」的 provider
            （GUI 步骤 3 / 语义检索置灰判据，spec §5.4 N2）。
        reason: 未就绪原因（供 GUI 定位引导起始步骤 + 诊断）。
    """

    model_config = {"from_attributes": True}

    ready: bool
    has_chat_model: bool
    has_embedding_model: bool
    reason: ReadinessReason

File: frontend/packages/renderer/src/stores/modelReadiness.ts
/**
 * F60 首启模型就绪 store（#934 §5.1/§5.6）——GUI 门控与语义检索置灰的单一真相。
 *
 * 判据来源：`GET /api/v1/settings/model-readiness`（后端派生判据，spec §2.1）。
 * 失败（网络/500）→ readiness=null（按「未就绪」保守处理，不阻塞可重试）。
 *
 * #1218 自愈：查询不再是一次性的——`startPolling` / `stopPolling` 让 App 在**未就绪**
 * 期间持续重查（镜像 `useKernelStore` 的轮询模式），外部（CLI/API/另一窗口）完成配置后
 * 主 UI 自动放行，无需手动刷新/重启。
 */
import { create } from 'zustand';
import { apiFetch } from '../api/client';

export interface ModelReadiness {
  ready: boolean;
  has_chat_model: boolean;
  has_embedding_model: boolean;
  reason: 'ready' | 'no_provider' | 'no_chat_model' | 'no_key';
}

interface ModelReadinessState {
  readiness: ModelReadiness | null;
  loading: boolean;
  /** 查询就绪判据（GET /api/v1/settings/model-readiness）；失败 → readiness=null */
  load: () => Promise<void>;
  /** #1218：未就绪期间持续重查（幂等；就绪后自行停止） */
  startPolling: () => void;
  /** #1218：停止重查（卸载时调用，防定时器泄漏） */
  stopPolling: () => void;
}

/** #1218 未就绪重查间隔（与 useKernelStore.POLL_INTERVAL_MS 同量级：轮询是兜底，代价可忽略） */
export const READINESS_POLL_INTERVAL_MS = 5000;

//: #1218 模块级单例定时器（镜像 useKernelStore）——App 实例与测试共享，stopPolling 可外部复位
let timer: ReturnType<typeof setInterval> | null = null;

export const useModelReadinessStore = create<ModelReadinessState>((set, get) => ({
  readiness: null,
  loading: false,

  load: async () => {
    set({ loading: true });
    try {
      const data = await apiFetch<ModelReadiness>('/api/v1/settings/model-readiness');
      set({ readiness: data, loading: false });
    } catch {
      // spec §7 边界 #6：查询失败不阻塞 GUI——按未就绪处理，可重试
      set({ readiness: null, loading: false });
    }
  },

  startPolling: () => {
    if (timer !== null) return; // 幂等
    void get().load();
    timer = setInterval(() => {
      if (get().readiness?.ready === true) {
        get().stopPolling(); // 就绪即停：避免无谓请求
        return;
      }
      void get().load();
    }, READINESS_POLL_INTERVAL_MS);
  },

  stopPolling: () => {
    if (timer !== null) {
      clearInterval(timer);
      timer = null;
    }
  },
}));

/**
 * N2 判据：语义检索是否可用（需 embedding 模型）。
 *
 * null（未查询/查询失败）→ false（保守：不可用）。
 */
export function canUseSemanticSearch(readiness: ModelReadiness | null): boolean {
  return readiness?.has_embedding_model === true;
}

/** 引导/主路径共用的模型管理入口 hash 路由（R5「前往配置」跳转目标）。 */
export const MODEL_SETTINGS_ROUTE = '/settings?cat=models';
